certora/specs/SocialRecoveryModule.spec

```
@@ -1,26 +1,209 @@
+    bool isReverted = lastReverted;
+
+    assert isReverted =>
+        isGuardian ||
+        e.msg.sender != safeContract ||
+        e.msg.value != 0 ||
+        guardian == 0 ||
+        guardian == SENTINEL() ||
+        guardian == safeContract ||
+        safeContract.isOwner(guardian) ||
+        threshold == 0 ||
+        to_mathint(threshold) > guardianStorageContract.entries[safeContract].count + 1 ||
+        guardianStorageContract.entries[safeContract].count == max_uint256 ||
+        !safeContract.isModuleEnabled(currentContract);
+}
+
+// This integrity rule verifies that if the revokeGuardianWithThreshold(...) executes, then ensure that:
+// - the Social Recovery Module is enabled
+// - the caller to the Module has to be the Safe Contract
+// - the guardian is revoked from the guardian list
+// - the linked list integrity remains,
+// - and no other account (guardian or not) is affected.
+rule revokeGuardiansWorksAsExpected(env e, address guardian, address prevGuardian, uint256 threshold, address otherAccount) {
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    address nextGuardian = guardianStorageContract.entries[safeContract].guardians[guardian];
+    bool otherAccountIsGuardian = currentContract.isGuardian(safeContract, otherAccount);
+
+    uint256 currentGuardiansCount = guardianStorageContract.entries[safeContract].count;
 
+    currentContract.revokeGuardianWithThreshold(e, safeContract, prevGuardian, guardian, threshold);
+
+    assert safeContract.isModuleEnabled(currentContract);
+    assert e.msg.sender == safeContract;
+    assert !currentContract.isGuardian(safeContract, guardian);
+    assert guardianStorageContract.entries[safeContract].guardians[prevGuardian] == nextGuardian;
+    assert guardian != otherAccount => otherAccountIsGuardian == currentContract.isGuardian(safeContract, otherAccount);
+    assert currentGuardiansCount > 0 => currentGuardiansCount - 1 == to_mathint(guardianStorageContract.entries[safeContract].count);
+    assert threshold <= guardianStorageContract.entries[safeContract].count;
+}
+
+// This integrity rule verifies that the guardian can always be revoked considering ideal conditions.
+rule guardianCanAlwaysBeRevoked(env e, address guardian, address prevGuardian, uint256 threshold) {
     requireSocialRecoveryModuleEnabled();
+    requireGuardiansLinkedListIntegrity(guardian);
+
+    // No value should be sent with the transaction.
+    require e.msg.value == 0;
+    // Guardian should not be zero address.
+    require guardian != 0;
+    // If new threshold is 0, then the guardian count should be 1.
+    require threshold == 0 => guardianStorageContract.entries[safeContract].count == 1;
+    // Else, if new threshold is greater than 0, then the guardian count should be greater than the threshold.
+    require threshold > 0 => guardianStorageContract.entries[safeContract].count > threshold;
```

```suggestion
    // The new threshold should be less than or equal to the guardian count after removing.
    require threshold <= guardianStorageContract.entries[safeContract].count - 1;
```

I don't think this needs a special case, the inequality holds for `threshold == 0` as well.

---

Right, updated 👍🏾